File: src/main/java/wav/demon/StatCraft/Commands/SC/SCXpGained.java
package wav.demon.StatCraft.Commands.SC;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import org.bukkit.command.CommandSender;
import wav.demon.StatCraft.Commands.ResponseBuilder;
import wav.demon.StatCraft.Querydsl.QPlayers;
import wav.demon.StatCraft.Querydsl.QXpGained;
import wav.demon.StatCraft.StatCraft;

import java.util.List;

public class SCXpGained extends SCTemplate {

    public SCXpGained(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("xpgained", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.xpgained");
    }

    @Override
    public String playerStatResponse(String name) {
        try {
            int id = plugin.getDatabaseManager().getPlayerId(name);
            if (id < 0)
                throw new Exception();

            SQLQuery query = plugin.getDatabaseManager().getNewQuery();
            QXpGained x = QXpGained.xpGained;
            Integer result = query.from(x).where(x.id.eq(id)).uniqueResult(x.amount);

            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Xp Gained")
                    .addStat("Total", df.format(result == null ? 0 : result))
                    .toString();
        } catch (Exception e) {
            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Xp Gained")
                    .addStat("Total", String.valueOf(0))
                    .toString();
        }
    }

    @Override
    public String serverStatListResponse(int num) {
        SQLQuery query = plugin.getDatabaseManager().getNewQuery();
        QXpGained x = QXpGained.xpGained;
        QPlayers p = QPlayers.players;

        List<Tuple> list = query
                .from(x)
                .leftJoin(p)
                .on(x.id.eq(p.id))
                .groupBy(p.name)
                .orderBy(x.amount.desc())
                .limit(num)
                .list(p.name, x.amount);

        return topListResponse("Xp Gained", list);
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCXpGained.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.commands.ResponseBuilder
import com.demonwav.statcraft.querydsl.QPlayers
import com.demonwav.statcraft.querydsl.QXpGained
import org.bukkit.command.CommandSender
import java.sql.Connection

class SCXpGained(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        plugin.baseCommand.registerCommand("xpgained", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<out String>?) = sender.hasPermission("statcraft.user.xpgained")

    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        val id = getId(name) ?: return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "Xp Gained" }
            stats["Total"] = "0"
        }

        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val x = QXpGained.xpGained
        val result = query.from(x).where(x.id.eq(id)).uniqueResult(x.amount.sum()) ?: 0

        return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "Xp Gained" }
            stats["Total"] = df.format(result)
        }
    }

    override fun serverStatListResponse(num: Long, args: List<String>, connection: Connection): String {
        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val x = QXpGained.xpGained
        val p = QPlayers.players

        val list = query
            .from(x)
            .innerJoin(p)
            .on(x.id.eq(p.id))
            .groupBy(p.name)
            .orderBy(x.amount.sum().desc())
            .limit(num)
            .list(p.name, x.amount.sum())

        return topListResponse("Xp Gained", list)
    }
}
